### DISCUSSION TITLE
timing out github action without "failure" status

### DISCUSSION THREAD

**Discussion by junefish** [community/community#38004] [Actions] [answered]
I am using Github Actions almost as a "cron job of sorts", to regularly run a Python script that syncs completed tasks from Todoist to Habitica.

When the [script](https://github.com/junefish/todoist-habitica-sync/blob/main/src/main.py) is run, either locally or with the [action YAML file](https://github.com/junefish/todoist-habitica-sync/blob/main/.github/workflows/run-command.yml), it keeps running (waiting for new completed tasks) until manually cancelled.  To prevent the action from running indefinitely, I added `timeout-minutes: 10` to the YAML file.

This works great in terms of the syncing behavior, but Github interprets the timeout as a "failure" (since technically the script isn't finishing successfully).  This generates a ton of noise in the form of a `Sync to Habitica workflow run failed for main branch` notification for every time the action runs.  I'd like to stop these so that a) my github notifications aren't cluttered and b) I can monitor if the run actually fails, e.g. if one of the APIs change.

TIA!

**Comment by antonblr** [comment, marked as answer]
I think a good pattern here is to actually instrument your application to run X minutes (could be configurable via env vars from CI) and then exit gracefully.

**Reply by junefish** [reply]
I'm new to CI and poetry and not sure how to configure this.  [pip has `PIP_DEFAULT_TIMEOUT`](https://pip-python3.readthedocs.io/en/latest/user_guide.html#environment-variables) but I don't see anything similar in the [poetry docs](https://python-poetry.org/docs/configuration/#available-settings).  Do you know how to do this?

Thanks!!

**Reply by antonblr** [reply]
I was talking about updating it on the code level. Basically, instead of [run_forever()](https://github.com/junefish/todoist-habitica-sync/blob/55c91d654129e28471ca1659770c9ae20021fc9b/src/main.py#L42) it should run `config.TIMEOUT || FOREVER` seconds (this is pseudo code). You can't do that with pip/poetry.

If you want something less involving you could look into killing/stopping it from the shell script. There is a `timeout` command from GNU core utils, but I'm not sure if it is available on GitHub Action runners. Test it (for error conditions as well):
```shell
timeout 600 poetry run python src/main.py`
```
This might work as well, but if `run_forever()` throw something other then `IOError` and exit, the job gonna wait X seconds anyways and then kill should exit with error code, since python process wouldn't exist anymore:
```shell
poetry run python src/main.py &
sleep 600
kill "$!"
```

**Reply by junefish** [reply]
Thank you, I was able to figure something out!

I removed `timeout-minutes: 10` from the YAML file and added the bolded lines to src/main.py:

<pre>
def run_forever(self):
        <b>timeout = time.time() + 60*10</b>
        while True:
            try:
                self._sync_todoist()
                self._next_tasks_state_based_on_todoist()
                self._next_tasks_state_in_habitica()
            except IOError as ex:
                self._log.error(f"Unexpected network error: {ex}")

            try:
                <b>if(time.time() < timeout):</b>
                    self._sync_sleep()
                <b>else:
                    break</b>
            except KeyboardInterrupt:
                break
</pre>

Maybe not the _most_ graceful, but the workflow run succeeded!
![screenshot](https://i.imgur.com/tmMncHG.png)

**Comment by airtower-luna** [comment]
I think the clean solution would be to add an option to the script to do a single-time sync. As in, sync the currently available items, then quit. Then you won't have to worry about a timeout at all, and the effect is about the same (after all after a timeout you'd also have to wait for the next run for new changes to sync). :slightly_smiling_face:

**Reply by junefish** [reply]
This makes sense but I think the delay + retry was setup by the OG author because of the Habitica API limitations.  It sometimes takes a few rounds of `(DelayTimer) [INFO]: Next check in 60 seconds` for it to finish syncing task updates.  So just taking out the delay+retry might not work.  (Yes, there's probably a way to check whether there are any active tasks and then exit, but I am an intermediate beginner :P.)

**Comment by gadimbaylisahil** [comment]
We also have another case which leads to similar annoyances.

- We open a PR or push a commit 
- This creates a new Github action
- Some of them are just ignored, and after 30 days, we get a timeout error.(none of the actions were approved though, they were never ran)

Any suggestions for this case?

**Comment by bertrandp** [comment]
You could add `continue-on-error: true` to have the desired ouput right ?

**Reply by anoopkkt** [reply]
Even with `continue-on-error: true` git marks it as failure, and blocks PRs from merging !!
